all/any builtins implemented using Truffle

OctaveLarose opened the pull request:
Hi all,

First of all, thank you for your work on this project! Second of all, while I've read the contributing guidelines and signed the OCA, I'm still waiting for it to be approved ; I thought I'd open this pull request to get some potential feedback in the meantime.

I've only been looking at this project for a couple of days and I'm not too familiar with it, just like I'm fairly new to Graal as well. Which is why I'm assuming this work should be revised.

While functional, it can be definitely improved in many ways I'm most likely not familiar with. As a series of notes:

- Unsure about a `while(true)` loop in the generic implementations for both. While I'm assuming a `StopIteration` exception will always be called and caught no matter what, I can't say it with full confidence out of lack of knowledge regarding GraalPython. Also unsure whether this implementation can prevent `KeyboardInterrupt`.
- No string specialization, as I assumed they weren't warranted as an `all` or `any` call on a `str` will always return `True` as far as I know anyway (unless `any` is fed an empty string). I could write a specialization method for strings that simply returns `true`, however. (and does a simple check for empty strings with `any`).
- I wrote a `dict` specialization but don't know how warranted it is for the reasons mentioned above. I made the `all` one return `True` by default because of them ; let me know if you can think of counterarguments to this implementation. (or this potential `PString` one)
- No `bytes`/`bytearray` specializations. I originally specialized using `PSequence`, which obviously caught them as they extend it, but I assumed specificity yielded better performance. Adding specifications for these types would be trivial.
- Small note: I match on `PSet` and not `PBaseSet`, hence not accounting for `PFrozenSet`. Unsure how relevant that is and how it should be done.
- All my code is in `BuiltinFunctions.java`, but I'd rather have the bulk of it be in one/two separate class(es), as seems to be the case for many builtin functions. However, I wasn't sure what to name them, where to put them in the file tree, etc. Which plays into my next question:
- I've been told specializing on different storage strategies may be pertinent, i.e subclasses of `SequenceStorage` instead of simply relying on this parent class (as I do here). If relevant, how should one go about this? This sounds like a lot of added logic if done right.

Thanks for your time and hard work!

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +296,154 @@ public Object absObject(VirtualFrame frame, Object object,
+                                            PyObjectIsTrueNode isTrueNode) {
+            SequenceStorage sequenceStorage = seq.getSequenceStorage();
+            Object[] internalArray = sequenceStorage.getInternalArray();
+            for (int i = 0; i < sequenceStorage.length(); i++)
```
Calling virtual methods like `sequenceStorage.length()` that have multiple implementations cannot always get inlined which has its performance cost. Under Truffle, we always try to avoid such virtual calls by making sure that the compiler has enough information to pick just one implementation, mostly by specializing on leaf types. For example, a few lines back you call `seq.getSequenceStorage()` and that is fine because this method will get inlined into the list (and tuple) specialization where the compiler knows that `seq`  is `PList` (`PTuple`) and the method in `PList` is `final`. But there's no such type information for the `length()` call here, the `sequenceStorage` could be any of the `SequenceStorages` subtypes.
For sequence storages, we have nodes that will try to specialize on the types such that virtual calls are avoided in most cases. For `length()`, there is `com.oracle.graal.python.builtins.objects.common.SequenceStorageNodes.LenNode`.
If you specialize on the storages as I outlined in the comment above, you can just call `length()` in specializations that specialize on a leaf type (i.e. `BooleanSequenceStorage`), but you will need `LenNode` if you have a generic fallback that covers more storages at the same time.

msimacek reviewed (commented):
Thank you for the contribution. The code is mostly correct, but has few performance problems, see the comments inline. I'll answer your questions in a spearate comment, the review UI is not good for writing long replies

OctaveLarose commented:
Thanks for your feedback! I'll try to send an updated version next week or so.

To answer a specific few comments:

> `dict` iteration yields the keys. There is no guarantee that all/any of them would be always true.

I did know it yielded the keys (which makes me realize calling `entries` was an obvious mistake I should have noticed), but actually had no idea they could be `None`... or anything other than strings, hence my default `true`. Frankly I don't know how I never came across that feature after years of using Python.

> I'd personally keep them in `BuiltinFunctions` to make them easy to find. They are not that big. Which "many builtin functions" do you see outside `BuiltinFunctions`?

I was just thinking of calls to various nodes, like `PyObjectSizeNode`, my wording was off. But really, my intent was to move at least part of the logic to a specialized node in the same way, which you've also recommended in a code comment, so that answers it.

> You may also consider generalizing the `all` and `any` operations into a common superclass or helper node that would have a boolean field or boolean argument that determines if the operation is `all` or `any`. 

I did consider it, actually! I thought it might be better to keep them separate for code clarity reasons and maybe performance (it's hard for me to assess what can impact performance negatively and what can't, for now). But yes, if so I'll gladly make them share some logic. Not a fan of boolean arguments and might use an `enum` instead for clarity, but that's hardly important.

Many questions I had but didn't necessarily ask got answered in your comments, so thank you for that! Thanks for taking the time to write these remarks, in general.

OctaveLarose commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -283,6 +291,224 @@ public Object absObject(VirtualFrame frame, Object object,
+         * Specializes on primitive storage types based on SequenceStorage.
+         */
+        @GenerateNodeFactory
+        public abstract static class AllOrAnySequenceStorageNode extends PythonUnaryBuiltinNode {
```
Fixed, it not having this parent type originally messed with the compilation for some reason, I believe... It may very well have been something else that was causing the issue, though, considering it works fine right now. I made it inherit from `PNodeWithContext`.

OctaveLarose commented:
That should be it, I thought I had more left to do - my bad for not getting it done last Friday. I wanted to try to improve the `@Specialization(limit = "3")` for `doHashStorage()` which I found a bit arbitrary, but I can't and I'm assuming there's a reason for the compiler to enforce this `limit` attribute being set for functions that rely on `@CachedLibrary` ; I could have set the limit to another value, but I preferred leaving it as is since I already don't understand it very well.

I also wanted to move `AllOrAnyNode` to another file/directory since as far as I know, `BuiltinFunctions.java` mostly only contains `@Builtin` annotated classes, but I suppose it wouldn't be the only exception to that rule.

In other news, Graal was indeed not up to date and `mx sforceimports` did the trick, I hadn't thought of that ; I had previously tried and failed with `mx clean`, though. Same for the benchmarks, I can now run them with `graalpython` ; I aim to get some working later this week for me to get an idea of what the speedup was.

msimacek commented:
Regarding benchmarks, there are two major metrics we measure:
- peak performance - the performance after the code has been running for some time and has been fully compiled
- interpreter performance - the performance without the graal compiler, which approximates how it performs during the "warm up"

We care a lot about peak performance because that's how we differentiate from CPython, but interpreter performance is also important so that programs don't take minutes to start up. Ideally, you would benchmark both. The default configuration measures peak performance. For benchmarking interpreter performance, you can run a benchmark like this: `mx --dy /compiler benchmark micro-small:name_of_your_benchmark -- --python-vm-config interpreter`.
AFAIK we don't have any benchmarks that would use `all` or `any`, so if you want to measure the improvements you did, you need to write your own benchmark (to run via `mx`, you need to add an entry for it in `mx.graalpython/mx_graalpython_bench_param.py`).

msimacek commented:
I ran the CI gates and there were some gate failures:
- JUnit tests for `all` failed. You can execute the failed suite using `mx punittest --no-leak-tests com.oracle.graal.python.test.builtin.BuiltinFunctionTests`. Since the builtins are now more complex, it might be a good idea to also add more tests to cover more specializations.
- The style gate failed due to formatting. Please run `mx python-style --fix --no-spotbugs`

OctaveLarose commented:
Fixed the tests, there was a big oversight in `all`'s behavior, sorry ; should be functional now (the tests do pass on my machine).

For the formatting, though, I'm not sure. `mx python-style --fix --no-spotbugs` doesn't seem to have turned up anything on my end and I can't see what I changed to the python code that could have broken it ; I did remove the `all` and `any` python functions, but I did make sure to keep the number of line breaks between the previous and next functions just the same as before. I also suspected the newline at the end of the file, but that's PEP 8 stuff and I don't believe I added one that wasn't already here before. Are CRLF/LF issues a possibility? I'm on Ubuntu LTS 20.04, if relevant.

> so if you want to measure the improvements you did, you need to write your own benchmark (to run via `mx`, you need to add an entry for it in `mx.graalpython/mx_graalpython_bench_param.py`).

Thanks! This should save me some searching.

OctaveLarose commented:
I rebased it on this repo's main branch again to make absolutely sure it wasn't the formatting issue didn't only happen when completely up to date, but this doesn't seem to have been the case.

msimacek commented:
Ah, sorry, I realized what the problem is with the formatter - it uses eclipse formatter and it needs to have eclipse installed, otherwise it only checks licenses. Please download eclipse (we use version 4.14 for the formatter, so it's probably a good idea to get the same), extract it somewhere and set environment variable `ECLIPSE_EXE` to the path of the `eclipse` binary. Then rerun the formatter.

Additionally, you can get the formatting in your IDE. If you're using Eclipse it should work already. If you're using Intellij IDEA you can install Eclipse Formatter Plugin and rerun `mx intellijinit`, it should configure itself.

OctaveLarose commented:
Ah, because of the name of the command, I was assuming it was purely for `.py` files. Anyway, I've ran the formatter using Eclipse v. 4.14 and should have fixed the issue

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +294,209 @@ public Object absObject(VirtualFrame frame, Object object,
+    /**
+     * Common class for all() and any() operations, as their logic and behaviors are very similar.
+     */
+    public abstract static class AllOrAnyNode extends PNodeWithContext {
```
Please make this package-private (not `public`)

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +294,209 @@ public Object absObject(VirtualFrame frame, Object object,
+    public abstract static class AllNode extends PythonUnaryBuiltinNode {
+
+        @Specialization(guards = "cannotBeOverridden(object, getClassNode)", limit = "1")
+        boolean doList(VirtualFrame frame,
```
The specialization methods here and in `AnyNode` could be `static`.

OctaveLarose replied:
Corrected in aa7cde4, but out of curiosity, is there a performance benefit to these methods being static? Or is it a code style/quality decision

msimacek replied:
There can be a small performance benefit in the interpreter for startup. In Java, invoking static methods can be a bit faster because there is no overriding, it doesn't have to lookup which method to call (there are two separate JVM instructions - `invokestatic` and `invokevirtual`). After it gets compiled by graal, there is no difference (the node types are constant and the method will get inlined, so there's no method call in the compiled code at all).

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +294,209 @@ public Object absObject(VirtualFrame frame, Object object,
+                        NodeType nodeType) {
+            boolean[] internalArray = sequenceStorage.getInternalBoolArray();
+
+            for (int i = 0; i < sequenceStorage.length(); i++) {
```
Optional: The loop condition (of all such `for` loops) could be profiled using `LoopConditionProfile`.

OctaveLarose replied:
Can an instance of `LoopConditionProfile` be kept in a field, or is it better for every specialization method to take one as parameter and feed it to the `AllOrAnyNode` methods?

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +294,209 @@ public Object absObject(VirtualFrame frame, Object object,
+                    if (!isTrueNode.execute(frame, next)) {
+                        return false;
+                    }
+                } catch (PException e) {
```
Optional: You could report the execution count of such `while(true)` loops using `LoopNode.reportLoopCount`.

OctaveLarose replied:
Would that not require making the `doObject` specialization methods not static, to be able to pass a `this` pointer to the method?

msimacek replied:
Yes, it would. Then don't make this one `static`. We just try to make methods `static` when they don't use `this` in any way, but it's not a problem at all to have them as instance methods when they do.

OctaveLarose replied:
Implemented in de15e38, with a call to it in `AllOrAnyNode.doHashStorage` as well (it used a `for` loop over keys of a `HashingStorage` object) ; it isn't a `while(true)` loop, and therefore I'm not sure if I should revert that specific change.

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +294,209 @@ public Object absObject(VirtualFrame frame, Object object,
+        }
+
+        @Specialization(guards = "cannotBeOverridden(object, getClassNode)", limit = "1")
+        boolean doSet(VirtualFrame frame,
```
The `doSet` and `doDict` specializations could be merged into one that takes `PHashingStorage`. `getDictStorage` is a `final` methods on `PHashingStorage`, so it's not necessary to specialize on the subclasses

msimacek reviewed (approved):
I forwarded some comments that I got from the internal review. There were some suggestions regarding loop profiling, which I marked as optional, because we're missing loop profiling in a lot of other places. Loop profiling is done to help the compiler have more accurate statistics about which code parts are "hotter" to guide other optimizations. I don't know relevant it is to your thesis, so I'm leaving it up to you if you want to deal with that.

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +297,204 @@ public Object absObject(VirtualFrame frame, Object object,
+                } else if (nodeType == NodeType.ANY && isTrueNode.execute(frame, internalArray[i])) {
+                    return true;
+                }
+            }
```
Missing `i++`? Btw, you can keep those as `for` loops, you can call the profile in the condition part (`for (int i = 0; i < loopConditionProfile.profile(i < seqLength); i++`)

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +297,204 @@ public Object absObject(VirtualFrame frame, Object object,
+                        NodeType nodeType,
+                        @CachedLibrary("hashingStorage") HashingStorageLibrary hlib) {
+            for (Object key : hlib.keys(hashingStorage)) {
+                LoopNode.reportLoopCount(this, 1);
```
You can profile this loop by converting it into a counted loop (`for (int i = 0; i < keys.length; i++`) and then profiling it like the other counted loops above.

msimacek commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +297,204 @@ public Object absObject(VirtualFrame frame, Object object,
+            Object iterator = getIter.execute(frame, object);
+            while (true) {
+                try {
+                    LoopNode.reportLoopCount(this, 1);
```
The reporting is supposed to be used to report the total number of iterations after the loop has finished. I'm not sure what happes when you report multiple times in a loop like this, but I think the compiler would simply average the reported numbers, so you would end up with the compiler assuming the loop is always executed once. What I believe you should do instead is to just count the number of iterations in a local variable and report it in a `finally` block as the last thing.

OctaveLarose replied:
Changed in bceb24a ; my bad for not reading the documentation closely enough for this one, from examples in the code I'd seen elsewhere, I had been assuming it aggregated the values fed for each `reportLoopCount` call in each scope (and I was really curious as to how that automatic detection worked...) 

Also, the infinite loops and better profiling for the (ex) for-each loop were fixed in 522e6ed.

steve-s commented on graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java:
```diff
@@ -286,6 +297,211 @@ public Object absObject(VirtualFrame frame, Object object,
+                        NodeType nodeType) {
+            boolean[] internalArray = sequenceStorage.getInternalBoolArray();
+
+            for (int i = 0; loopConditionProfile.profile(i < sequenceStorage.length()); i++) {
```
Thanks for adding the profiles. This is already higher standard than in most of our loops elsewhere in the code. For this specific case of counted loop, you can also use:
```
int len = sequenceStorage.length();
loopConditionProfile.profileCounted(len);
for (..., loopConditionProfile.inject(i < len), ...)
```
the `loopConditionProfile.profile` is incrementing a counter (look at the source), which are unnecessary instructions in the loop if we know the number of the iterations upfront. `inject` is intrinsic that only communicates the probability of branch taken/not taken to the compiler, but otherwise a no-op. One could argue that this is stretching it a bit too far, because the counter increment may be optimized away, but it is at least the best practice and a canonical example from the JavaDocs of `LoopConditionProfile`

OctaveLarose replied:
Sounds good, changed in 66661b4. Thanks for the explanation as well

steve-s replied:
Looks good now, thanks

msimacek commented:
You have a bug somewhere:
```
>>> any([False, True])
False
```

OctaveLarose commented:
> You have a bug somewhere:
> 
> ```
> >>> any([False, True])
> False
> ```

Thanks for the heads up, corrected in fa07d30 - was an oversight on my part.
